File: tests/ArrayListTest.php
<?php

use DSA\ArrayList\ArrayList;
use PHPUnit\Framework\TestCase;

class ArrayListTest extends TestCase
{
    public function test_append_and_get(): void
    {
        $list = new ArrayList;
        $list->append(1);
        $list->append(2);

        $this->assertSame(1, $list->get(0));
        $this->assertSame(2, $list->get(1));
    }

    public function test_size_and_capacity(): void
    {
        $list = new ArrayList;
        $this->assertSame(0, $list->size());
        $this->assertGreaterThanOrEqual(4, $list->capacity());
    }

    public function test_insert_and_prepend(): void
    {
        $list = new ArrayList;
        $list->append(1);
        $list->append(3);
        $list->insert(1, 2);
        $list->prepend(0);

        $this->assertSame(0, $list->get(0));
        $this->assertSame(1, $list->get(1));
        $this->assertSame(2, $list->get(2));
        $this->assertSame(3, $list->get(3));
    }

    public function test_delete_and_remove(): void
    {
        $list = new ArrayList;
        $list->append('a');
        $list->append('b');
        $list->append('c');

        $list->delete(1);
        $this->assertSame('a', $list->get(0));
        $this->assertSame('c', $list->get(1));
        $this->assertSame(2, $list->size());

        $list->remove('a');
        $this->assertSame('c', $list->get(0));
        $this->assertSame(1, $list->size());
    }

    public function test_remove_throws_on_missing(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $list = new ArrayList;
        $list->remove('not-found');
    }

    public function test_pop(): void
    {
        $list = new ArrayList;
        $list->append(42);
        $value = $list->pop();

        $this->assertSame(42, $value);
        $this->assertTrue($list->isEmpty());
    }

    public function test_pop_throws_when_empty(): void
    {
        $this->expectException(UnderflowException::class);
        $list = new ArrayList;
        $list->pop();
    }

    public function test_find(): void
    {
        $list = new ArrayList;
        $list->append('x');
        $list->append('y');
        $this->assertSame(1, $list->find('y'));
        $this->assertSame(-1, $list->find('z'));
    }

    public function test_resize(): void
    {
        $list = new ArrayList(2);
        $list->append(1);
        $list->append(2);
        // triggers resize
        $list->append(3);

        $this->assertSame(3, $list->size());
        $this->assertGreaterThanOrEqual(4, $list->capacity());
        $this->assertSame(3, $list->get(2));
    }

    public function test_set_and_get(): void
    {
        $list = new ArrayList;
        $list->append(1);
        $list->set(0, 10);
        $this->assertSame(10, $list->get(0));
    }

    public function test_set_throws_out_of_bounds(): void
    {
        $this->expectException(OutOfBoundsException::class);
        $list = new ArrayList;
        $list->set(0, 'fail');
    }

    public function test_get_throws_out_of_bounds(): void
    {
        $this->expectException(OutOfBoundsException::class);
        $list = new ArrayList;
        $list->get(0);
    }

    public function test_delete_throws_out_of_bounds(): void
    {
        $this->expectException(OutOfBoundsException::class);
        $list = new ArrayList;
        $list->delete(0);
    }
}

File: DSA/java/src/test/java/ds/ArrayListTest.java
package ds;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class ArrayListTest {

    @Test
    void testAppendAndGet() {
        ArrayList list = new ArrayList();
        list.append(1);
        list.append(2);

        assertEquals(1, list.get(0));
        assertEquals(2, list.get(1));
    }

    @Test
    void testSizeAndCapacity() {
        ArrayList list = new ArrayList();
        assertEquals(0, list.size());
        assertTrue(list.capacity() >= 4);
    }

    @Test
    void testInsertAndPrepend() {
        ArrayList list = new ArrayList();
        list.append(1);
        list.append(3);
        list.insert(1, 2); // between 1 and 3
        list.prepend(0); // at index 0

        assertEquals(0, list.get(0));
        assertEquals(1, list.get(1));
        assertEquals(2, list.get(2));
        assertEquals(3, list.get(3));
    }

    @Test
    void testDeleteAndRemove() {
        ArrayList list = new ArrayList();
        list.append("a");
        list.append("b");
        list.append("c");

        list.delete(1); // remove "b"
        assertEquals("a", list.get(0));
        assertEquals("c", list.get(1));
        assertEquals(2, list.size());

        list.remove("a");
        assertEquals("c", list.get(0));
        assertEquals(1, list.size());
    }

    @Test
    void testRemoveThrowsIfNotFound() {
        ArrayList list = new ArrayList();
        Exception e = assertThrows(IllegalArgumentException.class, () -> list.remove("x"));
        assertTrue(e.getMessage().contains("not found"));
    }

    @Test
    void testPop() {
        ArrayList list = new ArrayList();
        list.append(42);
        Object value = list.pop();

        assertEquals(42, value);
        assertTrue(list.isEmpty());
    }

    @Test
    void testPopThrowsIfEmpty() {
        ArrayList list = new ArrayList();
        assertThrows(IllegalStateException.class, list::pop);
    }

    @Test
    void testFind() {
        ArrayList list = new ArrayList();
        list.append("x");
        list.append("y");

        assertEquals(1, list.find("y"));
        assertEquals(-1, list.find("z"));
    }

    @Test
    void testResize() {
        ArrayList list = new ArrayList(2);
        list.append(1);
        list.append(2);
        list.append(3); // triggers resize

        assertEquals(3, list.size());
        assertTrue(list.capacity() >= 4);
        assertEquals(3, list.get(2));
    }

    @Test
    void testSetAndGet() {
        ArrayList list = new ArrayList();
        list.append(1);
        list.set(0, 10);
        assertEquals(10, list.get(0));
    }

    @Test
    void testSetThrowsOnInvalidIndex() {
        ArrayList list = new ArrayList();
        assertThrows(IndexOutOfBoundsException.class, () -> list.set(0, "fail"));
    }

    @Test
    void testGetThrowsOnInvalidIndex() {
        ArrayList list = new ArrayList();
        assertThrows(IndexOutOfBoundsException.class, () -> list.get(0));
    }

    @Test
    void testDeleteThrowsOnInvalidIndex() {
        ArrayList list = new ArrayList();
        assertThrows(IndexOutOfBoundsException.class, () -> list.delete(0));
    }
}
